File: Kth Ancestor in a Tree[26-July-2023]/Approach 2[recursive]/Solution.cpp
//  >> C++ CODE <<

int solve(Node* root, int k, int node, int* ans) {
    if (root == nullptr) {
        return -1;
    }
    if (root->data == node) {
        return 0;
    }
    
    int leftSearch = solve(root->left, k, node, ans);
    int rightSearch = solve(root->right, k, node, ans);
    
    if (leftSearch != -1 && rightSearch == -1) {
        if (leftSearch + 1 == k) {
            *ans = root->data;
        }
        return leftSearch + 1;
    } else if (rightSearch != -1 && leftSearch == -1) {
        if (rightSearch + 1 == k) {
            *ans = root->data;
        }
        return rightSearch + 1;
    }
    return -1;
}
int kthAncestor(Node* root, int k, int node) {
    int ans = -1;
    solve(root, k, node, &ans);
    return ans;
}

File: Kth Ancestor in a Tree[26-July-2023]/Approach 2[recursive]/solution.py
#  >> PYTHON CODE <<

def kthAncestor(root,k, node):
    def solve(root,k,node,ans):
        if(root==None): return -1
        if(root.data==node): return 0
        
        left_search=solve(root.left,k,node,ans)
        right_search=solve(root.right,k,node,ans)
        
        if(left_search!=-1 and right_search==-1):
            if(left_search+1==k):
                ans[0]=root.data
            
            return left_search+1
        elif(right_search!=-1 and left_search==-1):
            if(right_search+1==k):
                ans[0]=root.data
            
            return right_search+1
        
        return -1
        
    
    ans=[-1]
    solve(root,k,node,ans)

    return ans[0]